---
title: "McMurdo Dry Valleys Nutrient Limitation"
subtitle: "  \n Microbiome - 18S Taxonomic Summary"
author: "Adam J. Solon"
date: "`r Sys.Date()`"
#output: html_document
output: 
  pdf_document:
    toc: TRUE
    fig_width: 7
    fig_height: 6
    fig_caption: true
    keep_md: true
fontsize: 12pt
#editor_options: 
#  chunk_output_type: console
---
  
# Script Summary  
This script summarizes the taxonomic composition of photoautotrophs of the microeukaryotes (18S SSU rRNA gene), e.g., green algae, yellow-green algae, of samples from sediments collected in Taylor Vallye of the McMurdo Dry VAlleys.  The samples labelled 'bulk' are a homogenized mixtures of sediments collected in situ, while samples labelled 'CTRL', "N", "P", and "NP" are sediments exposed to a lab experiment with additions of nitrogen(N), phosphorus(P), both nutrients (NP), or neither (CTRL).

### Steps of this pipeline:  
### Steps of this pipeline:  
1.  Create and organize directories
2.  Load R packages
3.  Input files
4.  Format Files
5.  Phototroph summary - rarefied data
6. Save files and figures
  
```{r echo = FALSE, include = FALSE, set.seed(461)}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(include = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(cache = TRUE)

```

# Begin pipeline  
```{r, echo = FALSE, include = FALSE}
# Change identifiers to your system and file naming. 
user <- "F:"
directory <- "/Projects"
project <- "/CryoHoles/Studies"
study <- "/Soils-Nutrient-Limitation-Growth_Solon-et-al"

sub.directory.2 <- "/analyses"
sub.directory.2.1 <- "/microbiome"
sub.directory.2.1.1 <- "/amplicons"
sub.directory.2.1.1.1 <- "/18s"
sub.directory.2.1.1.1.1 <- "/input"
sub.directory.2.1.1.1.2 <- "/taxonomic_summary"
sub.directory.2.1.1.1.2.1 <- "/stats"
sub.directory.2.1.1.1.2.2 <- "/plots"

```

### Set pathways and create directories  

```{r set paths for pipeline}
# First define the project and project directories. 

# Create pathway for pipeline
###################################################
path.fp <- paste0(user, directory, project, study)
if (!dir.exists(path.fp)) dir.create(path.fp)

```

```{r set paths for analyses}
#sub.directory.2 analyses
###################################################
sub.directory.2.fp <- paste0(path.fp, sub.directory.2)
if (!dir.exists(sub.directory.2.fp)) dir.create(sub.directory.2.fp)

# Create sub-directory  microbiome
###################################################
sub.directory.2.1.fp <- paste0(sub.directory.2.fp, sub.directory.2.1)
if (!dir.exists(sub.directory.2.1.fp)) dir.create(sub.directory.2.1.fp)

# Create sub-directory amplicons
###################################################
sub.directory.2.1.1.fp <- paste0(sub.directory.2.1.fp, sub.directory.2.1.1)
if (!dir.exists(sub.directory.2.1.1.fp)) dir.create(sub.directory.2.1.1.fp)

# Create sub-directory 18s
###################################################
sub.directory.2.1.1.1.fp <- paste0(sub.directory.2.1.1.fp, sub.directory.2.1.1.1)
if (!dir.exists(sub.directory.2.1.1.1.fp)) dir.create(sub.directory.2.1.1.1.fp)

# Create sub-directory input
###################################################
sub.directory.2.1.1.1.1.fp <- paste0(sub.directory.2.1.1.1.fp, sub.directory.2.1.1.1.1)
if (!dir.exists(sub.directory.2.1.1.1.1.fp)) dir.create(sub.directory.2.1.1.1.1.fp)

# Create sub-directory beta diversity
###################################################
sub.directory.2.1.1.1.2.fp <- paste0(sub.directory.2.1.1.1.fp, sub.directory.2.1.1.1.2)
if (!dir.exists(sub.directory.2.1.1.1.2.fp)) dir.create(sub.directory.2.1.1.1.2.fp)

# Create sub-directory stats
###################################################
sub.directory.2.1.1.1.2.1.fp <- paste0(sub.directory.2.1.1.1.2.fp, sub.directory.2.1.1.1.2.1)
if (!dir.exists(sub.directory.2.1.1.1.2.1.fp)) dir.create(sub.directory.2.1.1.1.2.1.fp)

# Create sub-directory plots
###################################################
sub.directory.2.1.1.1.2.2.fp <- paste0(sub.directory.2.1.1.1.2.fp, sub.directory.2.1.1.1.2.2)
if (!dir.exists(sub.directory.2.1.1.1.2.2.fp)) dir.create(sub.directory.2.1.1.1.2.2.fp)

```

### Session Information  
```{r Install and load packages}
# install.packages("tidyverse")
# install.packages("knitr")
# install.packages("dplyr")
# install.packages("phyloseq")

library(knitr); packageVersion("knitr")
library(tidyverse); packageVersion("tidyverse")
library(dplyr); packageVersion("dplyr")
library(phyloseq); packageVersion("phyloseq")
library(patchwork)  # Load the patchwork package
library(gridExtra)
library(grid)

```  

* r version: `r getRversion()`
* RStudio version: `r rstudioapi::versionInfo()$version`
* r packages:  
  knitr, `r packageVersion("knitr")`  
  tidyverse, `r packageVersion("tidyverse")`  
  dplyr, `r packageVersion("dplyr")`  
  phyloseq, `r packageVersion("phyloseq")` 
  
### Input files
1. phyloseq object from r script - 'MDV-nutrient-limitation-soils-microbiome-normalization-transformation-18s.Rmd' 

```{r import phyloseq object}

#import phyloseq object with rarefaction of sequence counts (i.e., repeated random subsampling without replacement wiht same sequence depth for each sample) 
phy.r.2 <- readRDS(paste0(sub.directory.2.1.1.1.1.fp, "/phy.r.2.rds"))

```

### Plot COFily  
```{r stacked barplot cofily}
#plot using ggplot
#convert phyloseq object into data.frame
df.photo.r <- psmelt(phy.r.2)
  
# Remove rows where Abundance is 0
df.photo.r.1 <- df.photo.r %>%
  filter(Abundance != 0)

#subset by known photosynthetic eukaryotes, except for non-vascular plants (i.e., moss)
# Subset dataframe to keep only the desired divisions & remove "Spumella" genus
df.photo.r.2 <- df.photo.r.1 %>%
  filter(Division %in% c("Chlorophyta", "Ochrophyta") & Genus != "Spumella" & GenusSpecies != "Spumella-like_flagellate_JBAS36")

#"Dinoflagellata"
```

```{r format 2}
# Step 1: Reconstruct the OTU table
otu.tab <- df.photo.r.2 %>%
  select(Sample, OTU, Abundance) %>%
  spread(key = Sample, value = Abundance, fill = 0) %>%
  column_to_rownames("OTU")

# Convert into a phyloseq OTU table object
otu.tab.ps <- otu_table(as.matrix(otu.tab), taxa_are_rows = TRUE)

# Step 2: Reconstruct the Taxonomy table (Assuming columns for taxonomic ranks)
tax.tab <- df.photo.r.2 %>%
  select(OTU, Domain:GenusSpecies) %>%
  distinct() %>%
  column_to_rownames("OTU")

# Convert into a phyloseq Taxonomy table object
tax.tab.ps <- tax_table(as.matrix(tax.tab))

# Step 3: Reconstruct the Sample Metadata table
sam.tab <- df.photo.r.2 %>%
  select(Sample, X:dna_conc_ng_uL) %>%
  distinct() %>%
  column_to_rownames("Sample")

# Convert into a phyloseq Sample Data object
sam.tab.ps <- sample_data(sam.tab)

# Combine all components into a phyloseq object
phy.photo.r.3 <- phyloseq(otu.tab.ps, tax.tab.ps, sam.tab.ps)

```

```{r format phyloseq}
#transform rarefaction 'counts' into relative abundance percentage
phy.photo.r.4 <- transform_sample_counts(phy.photo.r.3, function(x) x / sum(x))

```

```{r}
#convert phyloseq object into data.frame
df.photo.r.3 <- psmelt(phy.photo.r.4)
  
#summarize relative abundance by cofily
df.photo.cof <- df.photo.r.3 %>%
  group_by(Sample, ClsOrdFam) %>%
  summarize(
    location = first(location),
    treatment = first(treatment),
    COFSum = sum(Abundance, na.rm = TRUE),
    .groups = "drop"
  )

#convert abundance into 0-100%
df.photo.cof$COFSum <- df.photo.cof$COFSum * 100


#set all low abundance as other
df.photo.cof$ClsOrdFam[df.photo.cof$COFSum < 1.0] <- "other"

#create a vector with all genera
vec.cof.photo <- unique(df.photo.cof$ClsOrdFam)

#set as dataframe
df.photo.cof.1 <- as.data.frame(df.photo.cof)

#reclassigy as factor
df.photo.cof.1 <- df.photo.cof.1 %>%
  mutate(ClsOrdFam = as.factor(ClsOrdFam))

# Reorder the 'location' factor
df.photo.cof.1$location <- factor(df.photo.cof.1$location, 
                              levels = c("Bonney", "Hoare", "Fryxell"))

# Reorder 'treatment'
df.photo.cof.1$treatment <- factor(df.photo.cof.1$treatment,
                               levels = c("bulk", "CTRL", "N", "P", "NP"))

#how many unique genera
num.cof.photo <- nlevels(df.photo.cof.1$ClsOrdFam)

#reorder factor levels of Phylum
df.photo.cof.1$ClsOrdFam <- factor(df.photo.cof.1$ClsOrdFam, levels=c("Bacillariophyceae",
                                                                      "Ochrophyta_Division",
                                                                      "Xanthophyceae",
                                                                      "Mischococcales",
                                                                      "Tribonematales",
                                                                      "Chlorophyta_Division",
                                                                      "Trebouxiophyceae",
                                                                      "Ulvophyceae",
                                                                      "Chlorophyceae",
                                                                      "Chlamydomonadales",
                                                                      "Chlorosarcinales",
                                                                      "Tetrasporales",
                                                                      "Chlorocystidales",
                                                                      "Chrysophyceae",
                                                                      "Ochromonadales",
                                                                      "Dinoflagellata_Division",
                                                                      "other"))

```
       
```{r plot COF}
#create facet labels
custom_facet_labels <- c(
  "Bonney" = "Bonney",
  "Hoare" = "Hoare",
  "Fryxell" = "Fryxell"
)

#create custom labels for taxa
custom_legend_labels <- c("Bacillariophyceae" = "Diatoms (Bacillariophyceae)",
                          "Ochrophyta_Division" = "Ochrophyta, other",
                          "Xanthophyceae" = "Yellow-Green Algae (Xanthophyceae)",
                          "Mischococcales" = "Yellow-Green Algae (Mischococcales)",
                          "Tribonematales" = "Yellow-Green Algae (Tribonematales)",
                          "Chlorophyta_Division" = "Green Algae, other",
                          "Trebouxiophyceae" = "Green Algae (Trebouxiophyceae)",
                          "Ulvophyceae" = "Green Algae (Ulvophyceae)",
                          "Chlorophyceae" = "Green Algae (Chlorophyceae)",
                          "Chlamydomonadales" = "Green Algae (Chlamydomonadales)",
                          "Chlorocystidales" = "Green Algae (Chlorocystidales)",
                          "Tetrasporales" = "Green Algae (Tetrasporales)",
                          "Chlorosarcinales" = "Green Algae (Chlorosarcinales)",
                          "Chrysophyceae" = "Golden Algae (Chrysophyceae)",
                          "Ochromonadales" = "Golden Algae (Ochromonadales)",
                          "Dinoflagellata_Division" = "Dinoflagellates",
                          "other" = "other")

#create a color palette
pal.1 <- c("#1E90FF", 
           '#F0E68C', "#FFD700", "#CDAD00", "#FFA500",
           "#006400", "#ADFF2F", "#00A000", "#00FF00", "#00C000", "#00E000", "#6B8E23", "#32CD32",
           "#FF0000", "#B22222",
           '#9A0EEA', 
           '#C5C9C7')

pal.2 <- c(
  "#0072B2", "#56B4E9", "#F0E68C", "#FFD700", "#CDAD00", "#FFA500",  # Blues & yellows (adjusted blues)
  "#1B9E77", "#66A61E", "#E6AB02", "#00FF00", "#6B8E23", "#32CD32",  # Greens & gold (better contrast)
  "#D81B60", "#E7298A", "#9A0EEA", "#FF69B4", "#C5C9C7"  # Magenta, pink, purple, gray + 1 added pink
)

# Create the bar plot
p.cof <- ggplot(data = df.photo.cof.1, aes(x = Sample, y = COFSum, fill = ClsOrdFam)) +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = pal.2, labels = custom_legend_labels) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(labels = function(x) {
    sapply(x, function(label) {
      # Extract the portion after the first digit and period using regular expressions
      sub(".*?(\\d+\\.\\d+)$", "\\1", label)
    })
  }) +
  labs(x = "Microcosms", y = "Relative Abundance (%)") +
  facet_grid(~ location + treatment, scales = "free_x", labeller = labeller(location = custom_facet_labels)) +
  theme_bw() +
  theme(
    legend.position = "bottom",
    legend.background = element_rect(fill = "grey90", colour = "grey50", size = 0.5),
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 8), # Adjust font size for x-axis tick marks
    strip.text.x = element_text(size = 10, face = "bold")
  ) +
  guides(fill = guide_legend(nrow = 5))

#display
p.cof

```

"#56B4E9",

```{r format cof 2}
#Closest BLAST match for the Ochrophyta_Division sequence is a genus of Diatom, Thalassiosira, at 86% identity. Change ambigious assignment to Bacillariophyceae (i.e., Diatoms)
df.photo.r.4 <- df.photo.r.3 %>%
  mutate(ClsOrdFam = ifelse(ClsOrdFam == "Ochrophyta_Division", "Bacillariophyceae", ClsOrdFam))

#summarize relative abundance by cofily
df.photo.2.cof <- df.photo.r.4 %>%
  group_by(Sample, ClsOrdFam) %>%
  summarize(
    location = first(location),
    treatment = first(treatment),
    COFSum = sum(Abundance, na.rm = TRUE),
    .groups = "drop"
  )

#convert abundance into 0-100%
df.photo.2.cof$COFSum <- df.photo.2.cof$COFSum * 100


#set all low abundance as other
df.photo.2.cof$ClsOrdFam[df.photo.2.cof$COFSum < 1.0] <- "other"

#create a vector with all genera
vec.2.cof.photo <- unique(df.photo.2.cof$ClsOrdFam)

#set as dataframe
df.photo.2.cof.1 <- as.data.frame(df.photo.2.cof)

#reclassigy as factor
df.photo.2.cof.1 <- df.photo.2.cof.1 %>%
  mutate(ClsOrdFam = as.factor(ClsOrdFam))

# Reorder the 'location' factor
df.photo.2.cof.1$location <- factor(df.photo.2.cof.1$location, 
                              levels = c("Bonney", "Hoare", "Fryxell"))

# Reorder 'treatment'
df.photo.2.cof.1$treatment <- factor(df.photo.2.cof.1$treatment,
                               levels = c("bulk", "CTRL", "N", "P", "NP"))

#how many unique genera
num.2.cof.photo <- nlevels(df.photo.2.cof.1$ClsOrdFam)

#reorder factor levels of Phylum
df.photo.2.cof.1$ClsOrdFam <- factor(df.photo.2.cof.1$ClsOrdFam, levels=c("Bacillariophyceae",
                                                                      #"Ochrophyta_Division",
                                                                      "Xanthophyceae",
                                                                      "Mischococcales",
                                                                      "Tribonematales",
                                                                      "Chlorophyta_Division",
                                                                      "Trebouxiophyceae",
                                                                      "Ulvophyceae",
                                                                      "Chlorophyceae",
                                                                      "Chlamydomonadales",
                                                                      "Chlorosarcinales",
                                                                      "Tetrasporales",
                                                                      "Chlorocystidales",
                                                                      "Chrysophyceae",
                                                                      "Ochromonadales",
                                                                      "Dinoflagellata_Division",
                                                                      "other"))

```
####

```{r format clades}
#convert phyloseq object into data.frame
df.photo.r.3 <- psmelt(phy.photo.r.4)

#new column grouping by common name clades
df.photo.r.4 <- df.photo.r.3 %>%
  mutate(clade = case_when(
    Division == "Chlorophyta" ~ "green_algae",
    Division == "Dinoflagellata" ~ "dinoflagellates",
    ClsOrdFam %in% c("Bacillariophyceae", "Ochrophyta_Division") ~ "diatoms",
    ClsOrdFam %in% c("Mischococcales", "Xanthophyceae", "Tribonematales") ~ "yellow-green_algae",
    ClsOrdFam %in% c("Chrysophyceae", "Ochromonadales") ~ "golden_algae",
    TRUE ~ NA_character_ # This ensures unmatched rows get NA
  ))

#summarize relative abundance by cldily
df.photo.cld <- df.photo.r.4 %>%
  group_by(Sample, clade) %>%
  summarize(
    location = first(location),
    treatment = first(treatment),
    CLDSum = sum(Abundance, na.rm = TRUE),
    .groups = "drop"
  )

#convert abundance into 0-100%
df.photo.cld$CLDSum <- df.photo.cld$CLDSum * 100


#set all low abundance as other
df.photo.cld$clade[df.photo.cld$CLDSum < 1.0] <- "other"

#create a vector with all genera
vec.cld.photo <- unique(df.photo.cld$clade)

#set as dataframe
df.photo.cld.1 <- as.data.frame(df.photo.cld)

#reclassigy as factor
df.photo.cld.1 <- df.photo.cld.1 %>%
  mutate(clade = as.factor(clade))

# Reorder the 'location' factor
df.photo.cld.1$location <- factor(df.photo.cld.1$location, 
                              levels = c("Bonney", "Hoare", "Fryxell"))

# Reorder 'treatment'
df.photo.cld.1$treatment <- factor(df.photo.cld.1$treatment,
                               levels = c("bulk", "CTRL", "N", "P", "NP"))

#how many unique genera
num.cld.photo <- nlevels(df.photo.cld.1$clade)

#reorder factor levels of Phylum
df.photo.cld.1$clade <- factor(df.photo.cld.1$clade, levels=c("diatoms",
                                                              "yellow-green_algae",
                                                              "green_algae",
                                                              "golden_algae",
                                                              "other"))

```

###keep unused factors of x-axis... scale_x_discrete(drop = FALSE)

```{r plot cld}
#create facet labels
facet_labels <- c(
  "Bonney" = "Bonney Basin",
  "Hoare" = "Hoare Basin",
  "Fryxell" = "Fryxell Basin"
)

#create custom labels for taxa
cld_legend_labels <- c("diatoms" = "Diatoms",
                          "green_algae" = "Green Algae",
                          "yellow-green_algae" = "Yellow-Green Algae",
                          "golden_algae" = "Golden Algae",
                          "other" = "other")

#create a color palette
pal.1 <- c("#0072B2", '#F0E68C', "#00C000", "#B8860B", '#C5C9C7')
   

# Create the bar plot
p.cld.face <- ggplot(data = df.photo.cld.1, aes(x = Sample, y = CLDSum, fill = clade)) +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = pal.1, labels = cld_legend_labels) +
  scale_y_continuous(expand = c(0,0)) +
  labs(x = "", y = "Relative Abundance (%)", fill = "Taxa") +
  guides(fill = guide_legend(nrow = 2)) +
  facet_wrap(~ location + treatment, nrow = 3, ncol = 5, scales = "free_x", labeller = labeller(location = facet_labels)) +
  theme_bw() +
  theme(legend.position = "bottom",
    legend.background = element_rect(fill = "grey90", colour = "grey50", size = 0.5),
    axis.text.x = element_blank(),
    strip.text.x = element_text(size = 10, face = "bold")
  )

#display
p.cld.face

```

```{r plot cld bn}
#subset for bonney
df.photo.cld.bn <- subset(df.photo.cld.1, location == "Bonney")

#add row for missing plate
row.1 = data.frame(Sample = "BNM_21", location = "Bonney", treatment = "P")
df.photo.cld.bn.1 <- bind_rows(df.photo.cld.bn, row.1)

#remove data from plate that was overwatered
df.photo.cld.bn.1 <- df.photo.cld.bn.1 %>%
  mutate(
    clade = if_else(Sample == "BNM_11", NA_character_, clade),
    CLDSum = if_else(Sample == "BNM_11", NA_real_, CLDSum)) %>%
  mutate(Sample = if_else(Sample == "BNM_11", "BNM_22", Sample))

# Reorder 'treatment'
df.photo.cld.bn.1$treatment <- factor(df.photo.cld.bn.1$treatment,
                               levels = c("bulk", "CTRL", "N", "P", "NP"))

#reorder factor levels of Phylum
df.photo.cld.bn.1$clade <- factor(df.photo.cld.bn.1$clade, levels=c("diatoms",
                                                              "yellow-green_algae",
                                                              "green_algae",
                                                              "golden_algae",
                                                              "other"))
#annotations for missing plates
df.anno <- data.frame(
  Sample = c("BNM_21", "BNM_22"),  # Replace with actual x-axis values
  CLDSum = c(10, 10),                 # Replace with desired y-axis positions
  treatment = c("P", "NP"), # Facet categories to appear in
  label = "x"
)

#treatment factor
df.anno$treatment <- factor(df.anno$treatment, levels = levels(df.photo.cld.bn.1$treatment))

#create a color palette
pal.bn <- c("#0072B2", '#F0E68C', "#00C000", "#B8860B", "grey50")
   
# Create the bar plot
p.cld.bn <- ggplot(data = df.photo.cld.bn.1, aes(x = Sample, y = CLDSum, fill = clade)) +
  geom_text(data = df.anno, aes(x = Sample, y = CLDSum, label = label), color = "black", size = 6, inherit.aes = FALSE) +  
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = pal.bn, labels = cld_legend_labels, na.translate = FALSE) +
  scale_y_continuous(expand = c(0,0)) +
  labs(title = "", subtitle = "Bonney Basin", fill = "Taxa") +
  guides(fill = guide_legend(nrow = 2)) +
  facet_wrap(~ treatment, ncol = 5, scales = "free_x") +
  theme_bw() +
  theme(plot.subtitle = element_text(size = 20),
          axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          axis.title.y = element_blank(),
          axis.text.y = element_text(size = 18),
          legend.position = "bottom",
          legend.background = element_rect(fill = "grey98", colour = "grey50", size = 0.5),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 14), 
          strip.text.x = element_text(size = 14),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.background = element_rect(fill = "grey95", colour = NA),
          axis.line = element_blank())

#display
p.cld.bn

#save R object
saveRDS(p.cld.bn, paste0(sub.directory.2.1.1.1.2.2.fp, "/p.cld.bn.rds"))

```

```{r plot cld hr}
#subset for hoare
df.photo.cld.hr <- subset(df.photo.cld.1, location == "Hoare")

#create a color palette
pal.hr <- c("#0072B2", '#F0E68C', "#00C000", '#C5C9C7')
 
# Create the bar plot
p.cld.hr <- ggplot(data = df.photo.cld.hr, aes(x = Sample, y = CLDSum, fill = clade)) +
  #geom_bar(stat = "identity", position = "stack", color = "black") +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = pal.hr, labels = cld_legend_labels) +
  scale_y_continuous(expand = c(0,0)) +
  labs(title = "Hoare Basin", x = "", y = "Relative Abundance (%)", fill = "Taxa") +
  guides(fill = guide_legend(nrow = 5)) +
  facet_wrap(~ treatment, ncol = 5, scales = "free_x") +
  theme_bw() +
  theme(plot.title = element_text(size = 20),
          axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          axis.title.y = element_text(size = 22),
          axis.text.y = element_text(size = 18),
          legend.position = "none",
          legend.background = element_rect(fill = "grey90", colour = "grey50", size = 0.5),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 14), 
          strip.text.x = element_text(size = 14)
  )

#display
p.cld.hr

#save R object
saveRDS(p.cld.hr, paste0(sub.directory.2.1.1.1.2.2.fp, "/p.cld.hr.rds"))

```

```{r plot cld fx}
#subset for fryxell
df.photo.cld.fx <- subset(df.photo.cld.1, location == "Fryxell")

# Define custom labels for the x-axis
fx_labels <- c(
  "FXM_01" = "C.1", "FXM_02" = "C.2", "FXM_06" = "C.3", "FXM_08" = "C.4", "FXM_14" = "C.5",
  "FXM_04" = "N.1", "FXM_11" = "N.2", "FXM_15" = "N.3", "FXM_16" = "N.4", "FXM_20" = "N.5",
  "FXM_05" = "P.1", "FXM_07" = "P.2", "FXM_09" = "P.3", "FXM_18" = "P.4", "FXM_19.2" = "P.5",
  "FXM_10" = "NP.1", "FXM_03" = "NP.2", "FXM_12" = "NP.3", "FXM_13" = "NP.4", "FXM_17" = "NP.5",
  "FX_Bulk_A" = "B.1", "FX_Bulk_B" = "B.2", "FX_Bulk_C" = "B.3", "FX_Bulk_D" = "B.4", 
  "FX_Bulk_E" = "B.5", "FX_Bulk_F" = "B.6"
)

#create a color palette
pal.fx <- c("#0072B2", '#F0E68C', "#00C000", "#B8860B", '#C5C9C7')

# Create the bar plot
p.cld.fx <- ggplot(data = df.photo.cld.fx, aes(x = Sample, y = CLDSum, fill = clade)) +
  #geom_bar(stat = "identity", position = "stack", color = "black") +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = pal.fx, labels = cld_legend_labels) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_discrete(labels = function(x) fx_labels[x]) + # Apply custom labels
  labs(title = "Fryxell Basin", x = "Microcosms", y = "", fill = "Taxa") +
  guides(fill = guide_legend(nrow = 5)) +
  facet_wrap(~ treatment, ncol = 5, scales = "free_x") +
  theme_bw() +
  theme(plot.title = element_text(size = 20),
          axis.title.x = element_text(size = 22),
          axis.text.x = element_text(size = 10),
          axis.title.y = element_blank(),
          axis.text.y = element_text(size = 18),
          legend.position = "none",
          legend.background = element_rect(fill = "grey90", colour = "grey50", size = 0.5),
          strip.text.x = element_text(size = 14)
  )

#display
p.cld.fx

#save R object
saveRDS(p.cld.fx, paste0(sub.directory.2.1.1.1.2.2.fp, "/p.cld.fx.rds"))

```

```{r}
# Extract the legend from bonney plot
legend <- ggplotGrob(p.cld.bn)$grobs[[which(sapply(ggplotGrob(p.cld.bn)$grobs, function(x) x$name) == "guide-box")]]

# Remove legend from bonney 
p.cld.bn.1 <- p.cld.bn + theme(legend.position = "none")

# Convert legend into a patchwork-compatible element
legend_patch <- wrap_elements(full = legend)

#combine plots
combined_plot <- (legend_patch / p.cld.bn.1 / p.cld.hr / p.cld.fx) + 
  plot_layout(heights = c(1, 3, 3, 3))  # Adjust the first item's height

#combined_plot <- (p.cld.bn.1 / p.cld.hr / p.cld.fx / legend_patch) + 
 # plot_layout(heights = c(3, 3, 3, 1))  # Adjust the last item's height

# Display the combined plot
print(combined_plot)

# Define file path for saving
file_path <- file.path(sub.directory.2.1.1.1.2.2.fp, "MDV-nutrient-limitation-soils-Taxa-Summary-18s-combined")

# Save combined plot in multiple formats
ggsave(filename = paste0(file_path, ".jpeg"), plot = combined_plot, width = 11, height = 8.5, dpi = 300)
ggsave(filename = paste0(file_path, ".pdf"), plot = combined_plot, width = 11, height = 8.5)
ggsave(filename = paste0(file_path, ".svg"), plot = combined_plot, width = 11, height = 8.5)

#save R object
saveRDS(combined_plot, paste0(sub.directory.2.1.1.1.2.2.fp, "/p.18s.rds"))

```

```{r combine plots}
# Modify each plot to remove extra margins and spacing
p.cld.bn.2 <- p.cld.bn.1 + theme(plot.margin = margin(0, 0, 1, 0))
p.cld.hr.1 <- p.cld.hr + theme(plot.margin = margin(1, 0, 0, 0))
p.cld.fx.1 <- p.cld.fx + theme(plot.margin = margin(1, 0, 0, 0))

# Combine the plots into one figure, each in its own row
p.cld.comb <- p.cld.bn.2 / p.cld.hr.1 / p.cld.fx.1 

# Display 
p.cld.comb

```

```{r save plots}
# Define parameters
file_names <- c("/MDV-nutrient-limitation-soils-taxa-summary-18s-facet", 
                "/MDV-nutrient-limitation-soils-taxa-summary-18s-no-legend")
plots <- list(p.cld.face, p.cld.comb)
formats <- c("jpeg", "pdf", "svg")

# Loop to save plots
for (i in seq_along(plots)) {
  for (format in formats) {
    ggsave(filename = paste0(sub.directory.2.1.1.1.2.2.fp, file_names[i], ".", format),
           plot = plots[[i]],
           device = format,
           width = 11,
           height = 8.5)
  }
}

```

```{r summary statistics}
#remove overwatered plate
df.photo.cld.2 <- df.photo.cld.1[df.photo.cld.1$Sample != "BNM_11", ]

#calculate mean and se
photo_cld_sum_stats <- df.photo.cld.2 %>%
  group_by(location, treatment, clade) %>%
  summarise(
    clade = first(clade),
    mean_CLDSum = round(mean(CLDSum, na.rm = TRUE), 1),  # Rounded to 1 decimal
    standard_error = round(ifelse(n() > 1, sd(CLDSum, na.rm = TRUE) / sqrt(n()), 0), 2),  # Rounded to 2 decimals
    n = n()  # Adds count column
  ) %>%
  ungroup()

# Display the updated dataframe
print(photo_cld_sum_stats)

# Save the dataframe as CSV
write.csv(photo_cld_sum_stats,  file.path(sub.directory.2.1.1.1.2.1.fp, paste0("photosynthetic_euks_clade_summary_stats.csv")), row.names = FALSE)

```

```{r inferential statistics}
# Load necessary libraries
library(car)  # For Levene's Test
library(nortest)  # For normality tests

# Open a connection to write output
sink(file.path(sub.directory.2.1.1.1.2.1.fp, "photosynthetic_euks_clade_inferential_stats.txt"))

# Loop through unique locations and clades
for (loc in unique(df.photo.cld.2$location)) {
  for (cld in unique(df.photo.cld.2$clade)) {
    
    # Skip clade "other"
    if (cld == "other") {
      cat("\nLocation:", loc, "- Skipping clade 'other'\n")
      next
    }
    
    # Subset data
    subset_data <- subset(df.photo.cld.2, location == loc & clade == cld)
    
    # Ensure there are enough observations
    if (nrow(subset_data) < 2) {
      cat("\nSkipping location:", loc, "clade:", cld, "- Not enough data\n")
      next
    }
    
    # Ensure treatment has at least two levels
    if (length(unique(subset_data$treatment)) < 2) {
      cat("\nSkipping location:", loc, "clade:", cld, "- Treatment has only one level\n")
      next
    }
    
    # Print location and clade
    cat("\nLocation:", loc, "Clade:", cld, "\n")
    
    # Perform ANOVA
    anova_model <- aov(CLDSum ~ treatment, data = subset_data)
    print(summary(anova_model))
    
    # Extract residuals
    residuals <- residuals(anova_model)
    
    # Test for normality
    shapiro_test <- shapiro.test(residuals)
    cat("\nShapiro-Wilk Normality Test:\n")
    print(shapiro_test)
    
    # Test for homogeneity of variance
    levene_test <- leveneTest(CLDSum ~ treatment, data = subset_data)
    cat("\nLevene's Test for Homogeneity of Variance:\n")
    print(levene_test)
    
    # Determine assumption violations
    normality_violated <- shapiro_test$p.value < 0.05
    variance_violated <- levene_test$Pr[1] < 0.05
    
    # If both assumptions are violated, print message only
    if (normality_violated & variance_violated) {
      cat("\nLocation:", loc, "Clade:", cld, "- Both assumptions are violated\n")
    } else {
      if (normality_violated) {
        cat("\nNormality assumption violated, using Kruskal-Wallis test\n")
        kruskal_test <- kruskal.test(CLDSum ~ treatment, data = subset_data)
        print(kruskal_test)
      }
      
      if (variance_violated) {
        cat("\nHomogeneity of variance assumption violated, using Welch's ANOVA\n")
        welch_anova <- oneway.test(CLDSum ~ treatment, data = subset_data, var.equal = FALSE)
        print(welch_anova)
      }
    }
  }
}

# Close the output file connection
sink()
```

```{r post hoc tukey}
# Load necessary library
library(multcomp)

# Open a connection to write output
sink(file.path(sub.directory.2.1.1.1.2.1.fp, "photosynthetic_euks_clade_post_hoc_tukey.txt"))

# Loop through unique locations and clades
for (loc in unique(df.photo.cld.2$location)) {
  for (cld in unique(df.photo.cld.2$clade)) {
    
    # Skip clade "other"
    if (cld == "other") {
      next
    }
    
    # Subset data
    subset_data <- subset(df.photo.cld.2, location == loc & clade == cld)
    
    # Ensure treatment has at least two levels
    if (length(unique(subset_data$treatment)) < 2) {
      next
    }

    # Perform ANOVA
    anova_model <- aov(CLDSum ~ treatment, data = subset_data)
    anova_results <- summary(anova_model)
    global_p_value <- anova_results[[1]]$`Pr(>F)`[1]

    # Test assumptions
    residuals <- residuals(anova_model)
    normality_test <- shapiro.test(residuals)
    variance_test <- leveneTest(CLDSum ~ treatment, data = subset_data)

    normality_met <- normality_test$p.value >= 0.05
    variance_met <- variance_test$Pr[1] >= 0.05

    # Perform Tukeyâ€™s test if ANOVA is significant and assumptions are met
    if (global_p_value < 0.05 & normality_met & variance_met) {
      cat("\nLocation:", loc, "Clade:", cld, "- Running Tukey HSD Test\n")
      tukey_test <- TukeyHSD(anova_model)
      print(tukey_test)
    }
  }
}

# Close the output file connection
sink()

```

```{r post hoc unn}
# Load necessary libraries
library(dunn.test)

# Open a connection to write output
sink(file.path(sub.directory.2.1.1.1.2.1.fp, "photosynthetic_euks_clade_post_hoc_dunn.txt"))

# Loop through unique locations and clades
for (loc in unique(df.photo.cld.2$location)) {
  for (cld in unique(df.photo.cld.2$clade)) {
    
    # Skip clade "other"
    if (cld == "other") {
      next
    }

    # Subset data
    subset_data <- subset(df.photo.cld.2, location == loc & clade == cld)

    # Ensure treatment has at least two levels
    if (length(unique(subset_data$treatment)) < 2) {
      next
    }

    # Perform Kruskal-Wallis Test only if assumptions were previously violated
    residuals <- residuals(aov(CLDSum ~ treatment, data = subset_data))
    normality_test <- shapiro.test(residuals)
    variance_test <- leveneTest(CLDSum ~ treatment, data = subset_data)

    normality_violated <- normality_test$p.value < 0.05
    variance_violated <- variance_test$Pr[1] < 0.05

    # Ensure Kruskal-Wallis was necessary in the first place
    if (normality_violated | variance_violated) {
      kruskal_test <- kruskal.test(CLDSum ~ treatment, data = subset_data)
      
      # Run Dunn's Test only if Kruskal-Wallis is significant
      if (kruskal_test$p.value < 0.05) {
        cat("\nLocation:", loc, "Clade:", cld, "- Running Dunn's Post Hoc Test\n")
        dunn_results <- dunn.test(subset_data$CLDSum, subset_data$treatment, method = "bonferroni")
        print(dunn_results)
      }
    }
  }
}

# Close the output file connection
sink()

```

## End of Script  
